src/hooks/useViewportColumns.ts

```diff
@@ -84,51 +78,84 @@ export function useViewportColumns<R, SR>({
+
+  const { layoutCssVars, totalColumnWidth, totalFrozenColumnWidth, columnMetrics } = useMemo(() => {
+    const columnMetrics = new Map<CalculatedColumn<R, SR>, ColumnMetric>();
+    let left = 0;
+    let totalColumnWidth = 0;
+    let totalFrozenColumnWidth = 0;
+    let templateColumns = '';
+    let allocatedWidth = 0;
+    let unassignedColumnsCount = 0;
+
+    for (const column of columns) {
+      let width = getSpecifiedWidth(column, columnWidths, viewportWidth);
+
+      if (width === undefined) {
+        unassignedColumnsCount++;
+      } else {
+        width = clampColumnWidth(width, column, minColumnWidth);
+        allocatedWidth += width;
+        columnMetrics.set(column, { width, left: 0 });
+      }
+    }
+
+    const unallocatedWidth = viewportWidth - allocatedWidth;
     const unallocatedColumnWidth = Math.max(
       Math.floor(unallocatedWidth / unassignedColumnsCount),
       minColumnWidth
     );
 
-    // Filter rawGroupBy and ignore keys that do not match the columns prop
-    const groupBy: string[] = [];
-    const calculatedColumns: CalculatedColumn<R, SR>[] = columns.map((column, idx) => {
-      // Every column should have a valid width as this stage
-      const width = column.width ?? clampColumnWidth(unallocatedColumnWidth, column, minColumnWidth);
-      const newColumn = {
-        ...column,
-        idx,
-        width,
-        left,
-        sortable: column.sortable ?? defaultSortable,
-        resizable: column.resizable ?? defaultResizable,
-        formatter: column.formatter ?? defaultFormatter
-      };
-
-      if (newColumn.rowGroup) {
-        groupBy.push(column.key);
-        newColumn.groupFormatter = column.groupFormatter ?? ToggleGroupFormatter;
+    for (const column of columns) {
+      let width;
+      if (columnMetrics.has(column)) {
+        const columnMetric = columnMetrics.get(column)!;
+        columnMetric.left = left;
+        ({ width } = columnMetric);
+      } else {
+        width = clampColumnWidth(unallocatedColumnWidth, column, minColumnWidth);
+        columnMetrics.set(column, { width, left });
       }
-
-      totalWidth += width;
+      totalColumnWidth += width;
       left += width;
-      return newColumn;
-    });
+      templateColumns += `${width}px `;
+    }
 
     if (lastFrozenColumnIndex !== -1) {
-      const lastFrozenColumn = calculatedColumns[lastFrozenColumnIndex];
-      lastFrozenColumn.isLastFrozenColumn = true;
-      totalFrozenColumnWidth = lastFrozenColumn.left + lastFrozenColumn.width;
+      const columnMetric = columnMetrics.get(columns[lastFrozenColumnIndex])!;
+      totalFrozenColumnWidth = columnMetric.left + columnMetric.width;
     }
 
-    return {
-      columns: calculatedColumns,
-      lastFrozenColumnIndex,
-      totalFrozenColumnWidth,
-      totalColumnWidth: totalWidth,
-      groupBy
+    const layoutCssVars: Record<string, string> = {
+      '--template-columns': templateColumns
```

Comment:
This is used for the `grid-template-columns` css property in all rows now.
That way we we the template on the root, and we don't have to update all rows/cells (in js).